File: go/functional/hash.go
package main

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"strings"
)

type HashFunc func(string) string

func generateSalt(length int) (string, error) {
	bytes := make([]byte, length)
	if _, err := rand.Read(bytes); err != nil {
		return "", err
	}
	return hex.EncodeToString(bytes), nil
}

func xorHasher(key byte) HashFunc {
	return func(input string) string {
		var hash strings.Builder
		for _, b := range []byte(input) {
			hashedByte := b ^ key
			hash.WriteString(fmt.Sprintf("%02x", hashedByte))
		}
		return hash.String()
	}
}

func main() {
	password := "password123"
	salt, err := generateSalt(16)
	if err != nil {
		panic(err)
	}

	saltedInput := password + salt
	hashFunc := xorHasher(0xAA)
	hash := hashFunc(saltedInput)

	fmt.Printf("Salt: %s\nHash: %s\n", salt, hash)
}

File: rust/functional/hash.rs
use rand::Rng;
use itertools::Itertools;

fn generate_salt(length: usize) -> String {
    let mut rng = rand::thread_rng();
    (0..length)
        .map(|_| format!("{:02x}", rng.r#gen::<u8>()))
        .collect()
}


fn simple_xor_hash(input: &str, key: u8) -> String {
    input.bytes()
        .map(|c| c ^ key)
        .map(|hashed_byte| format!("{:02x}", hashed_byte))
        .collect()
}

fn main() {
    let password = "password123";
    let salt = generate_salt(16);
    let salted_input = format!("{}{}", password, salt);
    let hash = simple_xor_hash(&salted_input, 0xAA);
    
    println!("Salt: {}\nHash: {}", salt, hash);
}
